File: cgi-bin/post.py
#!/usr/bin/env python

import os
import sys
import utils

try:
	path = utils.getEnvValue("ROOT_PATH") + ":" + utils.getEnvValue("PORT")
	os.makedirs(path, exist_ok=True)
	filename = utils.getEnvValue("UPLOAD_FILENAME")
	file_path = os.path.join(path, filename)
	status = 201
	status_message = "Created"
	message = ""
	
	if (os.path.isfile(file_path)):
		message = f"{filename} already exists."
		status = "400"
		status_message = "Bad Request"
	else:
		with open(file_path, "w") as w_file:
			for line in sys.stdin:
				w_file.write(line)
		message = f"{filename} has been posted successfully."
	
	
	dic_header = {
		"Connection" : utils.getEnvValue("CONNECTION"),
		"Content-Type" : utils.getEnvValue("CONTENT_TYPE"),
		"Content-Length" : len(message) + 1,
	}
	utils.printHeaderBody(dic_header, message, status, status_message)

except (utils.InternalServerError,  FileNotFoundError):
	with open("./page/error/500.html", "r") as file:
		err_page = file.read()
	dic_header = {
		"Connection" : utils.getEnvValue("CONNECTION"),
		"Content-Type" : utils.getEnvValue("CONTENT_TYPE"),
		"Content-Length" : len(err_page) + 1,
	}
	utils.printHeaderBody(dic_header, err_page, "500", "Internal Server Error")

File: cgi-bin/php/post.php
<?php

require_once("utils.php");

$path = getenv('ROOT_PATH') . ':' . getenv('PORT');

if (!file_exists($path)) {
	mkdir($path, 0764, true);
}

$file_path = $path . '/'. getenv('FILENAME_UPLOAD');
$status = 201;
$status_message = 'Created';
$message = '';

if (file_exists($file_path)) {
	$message = getenv('FILENAME_UPLOAD') . ' already exists';
	$status = 400;
	$status_message = 'Bad Request';
} else {
	$ofile = fopen($file_path, "w");
	while ($v = fgets(STDIN)) {
		$input .= $v;
	}
	$input = fgets(STDIN);
	fwrite($ofile, $input);
	fclose($ofile);
	$message = getenv('FILENAME_UPLOAD') . ' has been uploaded successfully.';
}

$dic_header = [
	'Connection' => getenv('CONNECTION'),
	'Content-Type' => getenv('CONTENT_TYPE'),
	'Content-Length' => strlen($message),
];
printHeaderBody($dic_header, $message, $status, $status_message);

?>